### ISSUE TITLE
Benchmarking scripts bench_gpu_1bn.py Error

### ISSUE THREAD

**Issue by zhaokang0826** [facebookresearch/faiss#252]
When I run the benchmarking scripts: bench_gpu_1bn.py like "python benchs/bench_gpu_1bn.py SIFT1M OPQ8_32,IVF262144,PQ8 -nnn 10 -ngpu 8 -tempmem $[1536*1024*1024]", it reports an error:
Traceback (most recent call last):
  File "benchs/bench_gpu_1bn.py", line 709, in <module>
    index = get_populated_index(preproc)
  File "benchs/bench_gpu_1bn.py", line 585, in get_populated_index
    gpu_index, indexall = compute_populated_index(preproc)
  File "benchs/bench_gpu_1bn.py", line 491, in compute_populated_index
    co.verbose = 10
  File "/gruntdata/baishao/faiss/swigfaiss_gpu.py", line 2331, in <lambda>
    __setattr__ = lambda self, name, value: _swig_setattr(self, GpuMultipleClonerOptions, name, value)
  File "/gruntdata/baishao/faiss/swigfaiss_gpu.py", line 51, in _swig_setattr
    return _swig_setattr_nondynamic(self,class_type,name,value,0)
  File "/gruntdata/baishao/faiss/swigfaiss_gpu.py", line 44, in _swig_setattr_nondynamic
    if method: return method(self,value)
TypeError: in method 'GpuClonerOptions_verbose_set', argument 2 of type 'bool'

It looks like the parameter co.verbose should be bool, so I change it to true or false. But both do not work:
Floating point exception

Can you help me?

Thank you!

**Comment by mdouze** [comment]
Hi 
Please replace `verbose = 10` with `verbose = True`. I changed it in the repo.

**Comment by zhaokang0826** [comment]
When I replace verbose = 10 with verbose = True in file bench_gpu_1bn.py line 491, I encounter another error:

**Comment by zhaokang0826** [comment]
...
making an IVFPQ index, m =  8
Training vector codes
  done 7.990 s
IndexShards shard 0 indices 0:0
**Floating point exception**
...

**Comment by zengzh** [comment]
I encountered this problem by using more than one GPU. There seems a problem sharding indexes.

**Comment by zhaokang0826** [comment]
When I use one GPU, I encountered the below error:
...
Faiss assertion 'err__ == cudaSuccess' failed in faiss::gpu::DeviceTensor<T, Dim, Contig, IndexT, PtrTraits>::~DeviceTensor() [with T = float; int Dim = 3; bool Contig = true; IndexT = int; PtrTraits = faiss::gpu::traits::DefaultPtrTraits] at impl/../utils/DeviceTensor-inl.cuh:61; details: CUDA error 77
...
Oh, my God

**Comment by zengzh** [comment]
This is due to the problem of precomputed tables. See issue  #250

**Comment by zengzh** [comment]
Hi @mdouze 

I tried to reproduce the experimetal results in the "GPU paper" by running the `bench_gpu_1bn.py` script under the `benchs `directory.

I used the SIFT1B dataset, and performed "IVF4096,PQ16" on smaller SIFT1M dataset to validate that everything is working.

First, I got the following error:

```
python benchs/bench_gpu_1bn.py SIFT1M IVF4096,PQ16 -nnn 10
...
store centroids /tmp/bench_gpu_1bn/cent_SIFT1M_IVF4096.npy
making an IVFPQ index, m =  16
Training vector codes
  done 128.690 s
IndexShards shard 0 indices 0:0
Floating point exception (core dumped)
```

From the output `IndexShards shard 0 indices **0:0**`, I figured out that there may be a problem of sharding indices for multiple GPUs. So I restricted the number of GPU to 1, using `-ngpu 1`. Then I got the error:

```
python benchs/bench_gpu_1bn.py SIFT1M IVF4096,PQ16 -nnn 10 -ngpu 1
...
Faiss assertion 'err__ == cudaSuccess' failed in faiss::gpu::DeviceTensor<T, Dim, Contig, IndexT, PtrTraits>::~DeviceTensor() [with T = float; int Dim = 3; bool Contig = true; IndexT = int; PtrTraits = faiss::gpu::traits::DefaultPtrTraits] at impl/../utils/DeviceTensor-inl.cuh:61; details: CUDA error 77
Aborted (core dumped)
```

I looked around on Github issues, and found out this is due to the problem of `precomputed tables` (issue #250 ). I tried to avoided this bug by commenting the `precompute_table()` function in `IndexIVFPQ.cpp`:

```
356	void IndexIVFPQ::precompute_table ()
357	{
358	    return;
359	}
```

After that, I got another error running the script:

```
...
Aggregate indexes to CPU
type of gpu_index: <class 'swigfaiss_gpu.GpuIndexIVFPQ'>
Traceback (most recent call last):
  File "benchs/bench_gpu_1bn.py", line 745, in <module>
    index = get_populated_index(preproc)
  File "benchs/bench_gpu_1bn.py", line 617, in get_populated_index
    gpu_index, indexall = compute_populated_index(preproc)
  File "benchs/bench_gpu_1bn.py", line 556, in compute_populated_index
    index_src = faiss.index_gpu_to_cpu(gpu_index.at(i))
  File "/home/zengzh/sdb1/faiss-master/swigfaiss_gpu.py", line 2563, in <lambda>
    __getattr__ = lambda self, name: _swig_getattr(self, GpuIndexIVFPQ, name)
  File "/home/zengzh/sdb1/faiss-master/swigfaiss_gpu.py", line 57, in _swig_getattr
    raise AttributeError(name)
AttributeError: at
```

This seems that `gpu_index` is expected to be a vector, but python doesn't recognize that. 

To sum up, there may exist several problems in faiss code:

1. sharding index for multiple GPUs
2. index_cpu_to_gpu when using precomputed tables
3. access gpu index from a vector in python script

Could you give some ideas to solve the above problems? Many thanks.

**Comment by mdouze** [comment]
I will push out a fix for `bench_bgpu_1bn` today or tomorrow.

**Comment by mdouze** [comment]
Should be fixed in latest version.

**Comment by reshu-b7** [comment]
Hi @mdouze ,

I have been using the most recent version of faiss library installed through conda.
While running the `bench_gpu_1bn.py` script I am still getting the above discussed bug.

```
Aggregate indexes to CPU
Traceback (most recent call last):
  File "bench_gpu_1bn.py", line 746, in <module>
    index = get_populated_index(preproc)
  File "bench_gpu_1bn.py", line 618, in get_populated_index
    gpu_index, indexall = compute_populated_index(preproc)
  File "bench_gpu_1bn.py", line 557, in compute_populated_index
    index_src = faiss.index_gpu_to_cpu(gpu_index.at(i))
  File "/opt/conda/miniconda2/lib/python2.7/site-packages/faiss/swigfaiss_gpu.py", line 4574, in <lambda>
    __getattr__ = lambda self, name: _swig_getattr(self, GpuIndexIVFPQ, name)
  File "/opt/conda/miniconda2/lib/python2.7/site-packages/faiss/swigfaiss_gpu.py", line 80, in _swig_getattr
    raise AttributeError("'%s' object has no attribute '%s'" % (class_type.__name__, name))
AttributeError: 'GpuIndexIVFPQ' object has no attribute 'at'
```

Has this not been fixed in the most recent version ??

**Comment by mdouze** [comment]
Hi 
It may be that the script does not work with a single GPU. Let me try.

**Comment by mdouze** [comment]
Fixed with:
```diff
diff --git a/benchs/bench_gpu_1bn.py b/benchs/bench_gpu_1bn.py
index 5b0bdd8..101d21d 100644
--- a/benchs/bench_gpu_1bn.py
+++ b/benchs/bench_gpu_1bn.py
@@ -552,7 +552,10 @@ def compute_populated_index(preproc):
     t0 = time.time()

     for i in range(ngpu):
-        index_src = faiss.index_gpu_to_cpu(gpu_index.at(i))
+        if ngpu == 1:
+            index_src = faiss.index_gpu_to_cpu(gpu_index)
+        else:
+            index_src = faiss.index_gpu_to_cpu(gpu_index.at(i))
         print "  index %d size %d" % (i, index_src.ntotal)
         index_src.copy_subset_to(indexall, 0, 0, nb)
```
Pushed to master.
